src/ssl.c

```diff
@@ -23360,6 +23360,192 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
+        {"3DES",                    NID_des_ede3_cbc},
+        {"CHACHA20/POLY1305(256)",  NID_chacha20_poly1305},
+        {"None",                    NID_undef},
+        {"IDEA",                    NID_idea_cbc},
+        {"RABBIT",                  NID_undef},
+        {"HC128",                   NID_undef},
+        {NULL,                      NID_undef}
+    };
+    
+    const struct ciphernid* c;
+    const char* encStr;
+    char n[MAX_SEGMENTS][MAX_SEGMENT_SZ] = {{0}};
+    
+    WOLFSSL_ENTER("wolfSSL_CIPHER_get_cipher_nid");
+    
+    GetCipherSegment(cipher, n);
+    encStr = GetCipherEncStr(n);
+    
+    if (encStr != NULL) {
+        for(c = ciphernid_tbl; c->alg_name != NULL; c++) {
+            if (XSTRNCMP(c->alg_name, encStr, XSTRLEN(c->alg_name)) == 0) {
+                return c->nid;
+            }
+        }
+    }
+    
+    return NID_undef;
+}
+/* return digest NID corresponding to cipher sutie
+ * @param cipher a pointer to WOLFSSL_CIPHER
+ * return NID if found, NID_undef if not found
+ */
+int wolfSSL_CIPHER_get_digest_nid(const WOLFSSL_CIPHER* cipher)
+{
+    static const struct macnid {
+        const char* alg_name;
+        const int  nid;
+    } macnid_tbl[] = {
+        {"SHA1",    NID_sha1},
+        {"SHA256",  NID_sha256},
+        {"SHA384",  NID_sha384},
+        {NULL,      NID_undef}
+    };
+    
+    const struct macnid* mc;
+    const char* name;
+    const char* macStr;
+    char n[MAX_SEGMENTS][MAX_SEGMENT_SZ] = {{0}};
+    (void)name;
+    
+    WOLFSSL_ENTER("wolfSSL_CIPHER_get_digest_nid");
+    
+    name = GetCipherSegment(cipher, n);
+    /* in MD5 case, NID will be NID_md5 */
+    if (XSTRSTR(name, "MD5") != NULL) {
+        return NID_md5;
+    }
+    
+    macStr = GetCipherMacStr(n);
+    
+    if (macStr != NULL) {
+        for(mc = macnid_tbl; mc->alg_name != NULL; mc++) {
+            if (XSTRNCMP(mc->alg_name, macStr, XSTRLEN(mc->alg_name)) == 0) {
+                return mc->nid;
+            }
+        }
+    }
+    
+    return NID_undef;
+}
+/* return key exchange NID corresponding to cipher sutie
+ * @param cipher a pointer to WOLFSSL_CIPHER
+ * return NID if found, NID_undef if not found
+ */
+int wolfSSL_CIPHER_get_kx_nid(const WOLFSSL_CIPHER* cipher)
+{
+static const struct kxnid {
+        const char* name;
+        const int  nid;
+    } kxnid_table[] = {
+        {"ECDHEPSK",  NID_kx_ecdhe_psk},
+        {"ECDH",      NID_kx_ecdhe},
+        {"DHEPSK",    NID_kx_dhe_psk},
+        {"DH",        NID_kx_dhe},
+        {"RSAPSK",    NID_kx_rsa_psk},
+        {"SRP",       NID_kx_srp},
+        {"EDH",       NID_kx_dhe},
+        {"RSA",       NID_kx_rsa},
+        {NULL,        NID_undef}
+    };
+    
+    const struct kxnid* k;
+    const char* name;
+    const char* keaStr;
+    char n[MAX_SEGMENTS][MAX_SEGMENT_SZ] = {{0}};
+    (void)name;
+    
+    WOLFSSL_ENTER("wolfSSL_CIPHER_get_kx_nid");
+    
+    name = GetCipherSegment(cipher, n);
```

Comment:
Should check return value of GetCipherSegment()

Reply:
Added checking return value, thanks